src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -74,3 +79,375 @@ xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
     return -1;
   }
 }
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> m_axlf;
+  const axlf* m_top;
+
+  void init_axlf_handle()
+  {
+    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
+    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
+      throw std::runtime_error("Invalid xclbin");
+    m_top = tmp;
+  }
+
+public:
+  xclbin_impl() : m_top(nullptr) {}
```

Comment:
Is there a need for an empty xclbin_impl?  I don't think this should be allowed.
xrt::xclbin can be empty indeed, but that just means   std::shared_ptr<xclbin_impl> handle; is a nullptr.

Reply:
When implementing xclbin.handle with a nullptr in the default constructor, subsequent calls on the object will produce a segfault. To avoid such ambiguity for the developer, we designed the class to throw an exception.

i.e.
xrt::xclbin xclbin;
xclbin.get_xsa_name(); // segfaults if using xclbin.handle == nullptr

Reply:
Even though I think this is a user error, maybe we don't need to support an empty xclbin at all?   I  know I thought about it to be  assignable, but is that is a use-case we don't need to care about?